File: internal/checks/js_libs_known_vuln.go
package checks

import (
	"context"
	"fmt"
	"strings"

	"github.com/londonmax12/hyperz/internal/httpclient"
	"github.com/londonmax12/hyperz/internal/page"
	"github.com/londonmax12/hyperz/internal/scope"
	"golang.org/x/net/html"
)

type JSLibsKnownVuln struct{}

func (JSLibsKnownVuln) Name() string { return "js-libs-known-vuln" }

func (JSLibsKnownVuln) Level() Level { return LevelPassive }

func (c JSLibsKnownVuln) Run(ctx context.Context, client *httpclient.Client, _ *scope.Scope, p page.Page) ([]Finding, error) {
	snap, err := ensureResponse(ctx, client, p, 1024*1024) // 1MB max body
	if err != nil {
		return nil, err
	}

	// Only process HTML responses.
	if !isHTMLContentType(snap.Headers.Get("Content-Type")) {
		return nil, nil
	}

	if len(snap.Body) == 0 {
		return nil, nil
	}

	detectedLibs := extractLibraries(string(snap.Body))
	if len(detectedLibs) == 0 {
		return nil, nil
	}

	var findings []Finding
	for lib, details := range detectedLibs {
		if len(details.vulnerabilities) == 0 {
			// Library detected but no known vulnerabilities - info level.
			findings = append(findings, Finding{
				Check:       c.Name(),
				Target:      p.URL,
				URL:         p.URL,
				Severity:    SeverityInfo,
				Title:       fmt.Sprintf("detected JavaScript library: %s", lib),
				Detail:      fmt.Sprintf("script analysis detected %s version %s; no known vulnerabilities for this version", lib, details.version),
				CWE:         "CWE-200",
				OWASP:       "A05:2021 Security Misconfiguration",
				Remediation: "Ensure all JavaScript libraries are kept up-to-date. Monitor security advisories for the libraries used.",
				Evidence:    BuildEvidence("GET", p.URL, snap.Status, snap.Headers, ""),
				DedupeKey:   MakeKey(c.Name(), ScopeHost, p.URL, "lib:"+lib),
			})
		} else {
			// Library detected with known vulnerabilities - medium/low level.
			details.vulnerabilities = append([]string{}, details.vulnerabilities...)
			var detail string
			if len(details.vulnerabilities) == 1 {
				detail = fmt.Sprintf("detected %s version %s which has a known vulnerability: %s", lib, details.version, details.vulnerabilities[0])
			} else {
				detail = fmt.Sprintf("detected %s version %s which has known vulnerabilities: %s", lib, details.version, strings.Join(details.vulnerabilities, ", "))
			}

			findings = append(findings, Finding{
				Check:       c.Name(),
				Target:      p.URL,
				URL:         p.URL,
				Severity:    SeverityMedium,
				Title:       fmt.Sprintf("%s (version %s) detected with known vulnerabilities", lib, details.version),
				Detail:      detail,
				CWE:         "CWE-1104",
				OWASP:       "A06:2021 Vulnerable and Outdated Components",
				Remediation: fmt.Sprintf("Update %s to the latest stable version. Check the project's security advisory page for details on what vulnerabilities have been patched.", lib),
				Evidence:    BuildEvidence("GET", p.URL, snap.Status, snap.Headers, ""),
				DedupeKey:   MakeKey(c.Name(), ScopeHost, p.URL, "vuln:"+lib+":"+details.version),
			})
		}
	}

	return findings, nil
}

type libDetails struct {
	version          string
	vulnerabilities  []string
}

// extractLibraries parses HTML to find script tags and identifies known libraries.
func extractLibraries(htmlBody string) map[string]libDetails {
	doc, err := html.Parse(strings.NewReader(htmlBody))
	if err != nil {
		return nil
	}

	detected := make(map[string]libDetails)

	var walker func(*html.Node)
	walker = func(n *html.Node) {
		if n.Type == html.ElementNode && n.Data == "script" {
			for _, attr := range n.Attr {
				if attr.Key == "src" {
					lib := identifyLibrary(attr.Val)
					if lib != nil {
						detected[lib.name] = libDetails{
							version:         lib.extractVersion(attr.Val),
							vulnerabilities: lib.getVulnerabilities(lib.extractVersion(attr.Val)),
						}
					}
				}
			}
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			walker(c)
		}
	}

	walker(doc)
	return detected
}

func identifyLibrary(scriptSrc string) *libPattern {
	lowerSrc := strings.ToLower(scriptSrc)
	for i := range jsLibPatterns {
		if jsLibPatterns[i].pattern.MatchString(lowerSrc) {
			// Return a copy with the matched pattern so we can extract version
			lib := jsLibPatterns[i]
			return &lib
		}
	}
	return nil
}

func (lp *libPattern) extractVersion(scriptSrc string) string {
	matches := lp.pattern.FindStringSubmatch(strings.ToLower(scriptSrc))
	if len(matches) > 1 {
		return matches[1]
	}
	return "unknown"
}

func (lp *libPattern) getVulnerabilities(version string) []string {
	// Try exact match first
	if vulns, ok := lp.vulnerableVersions[version]; ok && len(vulns) > 0 {
		return vulns
	}
	// Try major.minor without patch version
	parts := strings.Split(version, ".")
	if len(parts) >= 2 {
		majorMinor := strings.Join(parts[:2], ".")
		if vulns, ok := lp.vulnerableVersions[majorMinor]; ok && len(vulns) > 0 {
			return vulns
		}
	}
	return nil
}

File: internal/checks/supply_chain/js_libs_known_vuln.lua
-- js-libs-known-vuln: scans <script src> tags on an HTML response,
-- identifies known JS libraries from their URL pattern, and emits one
-- finding per (library, version) pair. A version that matches a row
-- in the per-library vulnerable-versions table escalates to medium
-- with a CVE / vuln-class list; an unmatched version stays info.

local check = {
  name  = "js-libs-known-vuln",
  level = levels.passive,
  scope = scopes.host,
  tier  = tiers.fingerprint,
}

function check.run(ctx)
  local snap, err = ctx:ensure_response{ max_body = 1024 * 1024 }
  if err then return nil, err end

  -- HTML-only; non-HTML responses (JSON APIs, asset files) carry no
  -- <script src> markup for the scanner to walk.
  if not ctx.body.is_html_ct(snap.headers:get("Content-Type")) then
    return nil
  end
  if snap.body == "" then return nil end

  local hits = ctx.supply_chain.scan_known_js_libs(snap.body)
  if #hits == 0 then return nil end

  local evidence = ctx.evidence.build {
    method  = methods.get,
    url     = ctx.page.url,
    status  = snap.status,
    headers = snap.headers,
  }

  local findings = {}
  for _, h in ipairs(hits) do
    if #h.vulnerabilities == 0 then
      -- Library detected but no known vulns at this version - info.
      findings[#findings + 1] = {
        severity    = severity.info,
        title       = "detected JavaScript library: " .. h.name,
        detail      = string.format(
          "script analysis detected %s version %s; no known vulnerabilities for this version",
          h.name, h.version),
        cwe         = "CWE-200",
        owasp       = "A05:2021 Security Misconfiguration",
        remediation = "Ensure all JavaScript libraries are kept up-to-date. Monitor security advisories for the libraries used.",
        evidence    = evidence,
        dedupe_parts = { "lib:" .. h.name },
      }
    else
      -- Vulnerable version detected; the title carries the library +
      -- version so the report row distinguishes "jquery 1.7" from
      -- "jquery 2.1" without forcing the reader into Details.
      local detail
      if #h.vulnerabilities == 1 then
        detail = string.format(
          "detected %s version %s which has a known vulnerability: %s",
          h.name, h.version, h.vulnerabilities[1])
      else
        detail = string.format(
          "detected %s version %s which has known vulnerabilities: %s",
          h.name, h.version, table.concat(h.vulnerabilities, ", "))
      end
      findings[#findings + 1] = {
        severity    = severity.medium,
        title       = string.format("%s (version %s) detected with known vulnerabilities", h.name, h.version),
        detail      = detail,
        cwe         = "CWE-1104",
        owasp       = "A06:2021 Vulnerable and Outdated Components",
        remediation = string.format(
          "Update %s to the latest stable version. Check the project's security advisory page for details on what vulnerabilities have been patched.",
          h.name),
        evidence    = evidence,
        dedupe_parts = { "vuln:" .. h.name .. ":" .. h.version },
      }
    end
  end
  return findings
end

return check
